jsonpickle/handlers.py

```diff
@@ -196,11 +197,30 @@ def restore(self, data):
         return cls.__new__(cls, *params)
 
 
-DatetimeHandler.handles(datetime.datetime)
 DatetimeHandler.handles(datetime.date)
 DatetimeHandler.handles(datetime.time)
 
 
+class DatetimeObjectHandler(BaseHandler):
+    """Serialization for datetime.datetime objects"""
+
+    def flatten(self, obj, data):
+        if not self.context.unpicklable:
+            return obj.isoformat()
+        data['iso'] = obj.isoformat()
+        data['tzinfo'] = self.context.flatten(obj.tzinfo, reset=False)
+        return data
+
+    def restore(self, data):
+        obj = dateutil.parser.isoparse(data['iso'])
+        tzinfo = self.context.restore(data['tzinfo'], reset=False)
+        obj = obj.replace(tzinfo=tzinfo)
+        return obj
+
+
+DatetimeObjectHandler.handles(datetime.datetime)
```

Comment:
I'd be interested in what impact this has on subclasses of datetime. The `__reduce__` protocol allows subclasses to be serialized and restored and for those subclasses to supply the necessary context to flatten and restore subclass-specific details. I did some experimentation and determined that subclasses get handled by another handler:

```
draft $ cat jpformat.py                                                                                                                                                   
import datetime
import jsonpickle


class mydatetime(datetime.datetime):
    pass


pickle = jsonpickle.dumps(mydatetime.now())
print(pickle)
restored = jsonpickle.loads(pickle)
print(type(restored))
draft $ pip-run --use-pep517 git+https://github.com/davvid/jsonpickle@isoformat python-dateutil -- -m jpformat                                                            
  Running command git clone -q https://github.com/davvid/jsonpickle /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-req-build-vqw99xxk
Collecting git+https://github.com/davvid/jsonpickle@isoformat
  Cloning https://github.com/davvid/jsonpickle (to revision isoformat) to /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-req-build-vqw99xxk
  Running command git checkout -b isoformat --track origin/isoformat
  Switched to a new branch 'isoformat'
  Branch 'isoformat' set up to track remote branch 'isoformat' from 'origin'.
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting python-dateutil
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting six>=1.5
  Using cached six-1.14.0-py2.py3-none-any.whl (10 kB)
Building wheels for collected packages: jsonpickle
  Building wheel for jsonpickle (PEP 517) ... done
  Created wheel for jsonpickle: filename=jsonpickle-1.3-py2.py3-none-any.whl size=33829 sha256=54646a11b98ca82390ce01a37fc81cb35e007ef43871388561f888af567c2ee3
  Stored in directory: /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-ephem-wheel-cache-kjkzgfiv/wheels/ee/df/ad/8fd9553b7d085584a9c8a28a64c12c6eba29cc9b9daad96428
Successfully built jsonpickle
Installing collected packages: six, python-dateutil, jsonpickle
Successfully installed jsonpickle-1.3 python-dateutil-2.8.1 six-1.14.0
{"py/reduce": [{"py/type": "__main__.mydatetime"}, {"py/tuple": [{"py/b64": "B+QCFwohNwSlOw=="}]}]}
<class '__main__.mydatetime'>
```